sunscreen_math/src/metal_impl/ristrettovec.rs

```diff
@@ -0,0 +1,716 @@
+    /// ```rust
+    /// struct RistrettoPoint(EdwardsPoint);
+    ///
+    /// struct EdwardsPoint {
+    ///     X: FieldElement2625,
+    ///     Y: FieldElement2625,
+    ///     Z: FieldElement2625,
+    ///     T: FieldElement2625,
+    /// }
+    ///
+    /// struct FieldElement2625([u32; 10]);
+    /// ```
+    /// To achieve this layout, you must use the u32 backend.
+    ///
+    pub fn new(x: &[RistrettoPoint]) -> Self {
+        let runtime = Runtime::get();
+
+        let len = x.len();
+
+        assert_eq!(size_of::<RistrettoPoint>(), size_of::<u32>() * 40);
+        let byte_len = x.len() * size_of::<RistrettoPoint>();
+        let data = runtime.alloc(byte_len);
+
+        let mut field_vec = Self { data, len };
+
+        let data_slice = field_vec.buffer_slice_mut();
+
+        for (i, p) in x.iter().enumerate() {
+            let x = p.0.X.to_u29();
+            let y = p.0.Y.to_u29();
+            let z = p.0.Z.to_u29();
+            let t = p.0.T.to_u29();
+
+            for (j, w) in x.iter().enumerate() {
+                data_slice[(j) * len + i] = *w;
+            }
+
+            for (j, w) in y.iter().enumerate() {
+                data_slice[(j + 10) * len + i] = *w;
+            }
+
+            for (j, w) in z.iter().enumerate() {
+                data_slice[(j + 20) * len + i] = *w;
+            }
+
+            for (j, w) in t.iter().enumerate() {
+                data_slice[(j + 30) * len + i] = *w;
+            }
+        }
+
+        field_vec
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn len_bytes(&self) -> usize {
+        self.len * size_of::<RistrettoPoint>()
+    }
+
+    pub fn get(&self, index: usize) -> RistrettoPoint {
+        if index > self.len {
+            panic!("Index {index} exceeds bounds of {}", self.len);
+        }
+
+        let mut x = [0u32; 10];
+        let mut y = [0u32; 10];
+        let mut z = [0u32; 10];
+        let mut t = [0u32; 10];
+
+        let buffer_slice = self.buffer_slice();
+
+        for i in 0..10 {
+            x[i] = buffer_slice[i * self.len + index];
+        }
+
+        for i in 0..10 {
+            y[i] = buffer_slice[(i + 10) * self.len + index];
+        }
+
+        for i in 0..10 {
+            z[i] = buffer_slice[(i + 20) * self.len + index];
+        }
+
+        for i in 0..10 {
+            t[i] = buffer_slice[(i + 30) * self.len + index];
+        }
+
+        RistrettoPoint(EdwardsPoint {
+            X: CannonicalFieldElement(x).to_field(),
+            Y: CannonicalFieldElement(y).to_field(),
+            Z: CannonicalFieldElement(z).to_field(),
+            T: CannonicalFieldElement(t).to_field(),
+        })
+    }
+
+    pub fn iter(&self) -> RistrettoPoints {
+        RistrettoPoints {
+            vec: self,
+            i: 0usize,
+        }
+    }
+
+    // TODO: This probably needs to return a slice of MaybeUninit<u32>.
+    fn buffer_slice_mut(&mut self) -> &mut [u32] {
```

Comment:
Do you want to do this now? Or is this comment not true anymore?

Reply:
I don't actually know the rules around FFI and uninitialized memory and need to study this a bit more. Accessing uninit data in Rust is UB, but if FFI returned a region of memory it allocated, how would Rust tell the difference? I think I just want to leave the TODO here for now and I'll make an Issue for someone to research and come up with the answer.